File: rust/business-logic/classic-message-core/src/enums.rs
//! Message type and target enumerations for the MessageHandler system.
use serde::{Deserialize, Serialize};

/// Represents various types of messages categorized by their context and severity.
///
/// This enumeration helps to define and manage message categories efficiently
/// for logging or displaying purposes. It provides a predefined set of message
/// types that can be used across an application to standardize how messages are
/// identified and processed.
///
/// # Examples
///
/// ```rust
/// use classic_message_core::MessageType;
///
/// let msg_type = MessageType::Info;
/// assert_eq!(msg_type.to_log_level(), log::Level::Info);
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum MessageType {
    /// Denotes an informational message.
    Info,
    /// Denotes a warning message indicating a potential issue.
    Warning,
    /// Denotes an error message indicating a failure or problem.
    Error,
    /// Denotes a message indicating the successful completion of an operation or process.
    Success,
    /// Denotes a message signaling the progress of an ongoing operation.
    Progress,
    /// Denotes a message intended for debugging purposes.
    Debug,
    /// Denotes a message indicating a critical issue requiring immediate attention.
    Critical,
}

impl MessageType {
    /// Converts the message type to its corresponding log level.
    ///
    /// Maps message types to standard logging levels for consistent logging behavior:
    /// - Info, Success → log::Level::Info
    /// - Warning → log::Level::Warn
    /// - Error, Critical → log::Level::Error
    /// - Debug, Progress → log::Level::Debug
    ///
    /// # Returns
    ///
    /// The corresponding `log::Level` for this message type.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageType;
    ///
    /// assert_eq!(MessageType::Error.to_log_level(), log::Level::Error);
    /// assert_eq!(MessageType::Info.to_log_level(), log::Level::Info);
    /// ```
    #[must_use]
    pub const fn to_log_level(&self) -> log::Level {
        match self {
            Self::Debug | Self::Progress => log::Level::Debug,
            Self::Info | Self::Success => log::Level::Info,
            Self::Warning => log::Level::Warn,
            Self::Error | Self::Critical => log::Level::Error,
        }
    }

    /// Returns a human-readable name for the message type.
    ///
    /// # Returns
    ///
    /// A string slice representing the name of the message type.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageType;
    ///
    /// assert_eq!(MessageType::Info.name(), "Info");
    /// assert_eq!(MessageType::Critical.name(), "Critical");
    /// ```
    #[must_use]
    pub const fn name(&self) -> &'static str {
        match self {
            Self::Info => "Info",
            Self::Warning => "Warning",
            Self::Error => "Error",
            Self::Success => "Success",
            Self::Progress => "Progress",
            Self::Debug => "Debug",
            Self::Critical => "Critical",
        }
    }
}

/// Enumeration to specify where a message should be directed.
///
/// This enum defines various message targets such as both GUI and CLI,
/// only GUI, only CLI, or logs. It is used to control and direct how
/// messages are processed and displayed in an application context.
///
/// # Examples
///
/// ```rust
/// use classic_message_core::MessageTarget;
///
/// let target = MessageTarget::All;
/// assert!(target.should_display_in_gui());
/// assert!(target.should_display_in_cli());
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
pub enum MessageTarget {
    /// Show in both GUI and CLI.
    #[default]
    All,
    /// Show only in GUI mode (legacy, prefer `Gui`).
    GuiOnly,
    /// Show only in CLI mode (legacy, prefer `Console`).
    CliOnly,
    /// Only write to log file, no display.
    LogOnly,
    /// Show only in GUI mode.
    Gui,
    /// Show only in CLI mode.
    Console,
}

impl MessageTarget {
    /// Determines if the message should be displayed in GUI mode.
    ///
    /// # Arguments
    ///
    /// * `is_gui_mode` - Whether the application is currently in GUI mode.
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed in GUI, `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display_in_gui());
    /// assert!(MessageTarget::Gui.should_display_in_gui());
    /// assert!(!MessageTarget::Console.should_display_in_gui());
    /// assert!(!MessageTarget::LogOnly.should_display_in_gui());
    /// ```
    #[must_use]
    pub const fn should_display_in_gui(&self) -> bool {
        matches!(self, Self::All | Self::GuiOnly | Self::Gui)
    }

    /// Determines if the message should be displayed in CLI mode.
    ///
    /// # Arguments
    ///
    /// * `is_gui_mode` - Whether the application is currently in GUI mode.
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed in CLI, `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display_in_cli());
    /// assert!(MessageTarget::Console.should_display_in_cli());
    /// assert!(!MessageTarget::Gui.should_display_in_cli());
    /// assert!(!MessageTarget::LogOnly.should_display_in_cli());
    /// ```
    #[must_use]
    pub const fn should_display_in_cli(&self) -> bool {
        matches!(self, Self::All | Self::CliOnly | Self::Console)
    }

    /// Determines if the message should be displayed at all (not log-only).
    ///
    /// # Returns
    ///
    /// `true` if the message should be displayed (not log-only), `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use classic_message_core::MessageTarget;
    ///
    /// assert!(MessageTarget::All.should_display());
    /// assert!(!MessageTarget::LogOnly.should_display());
    /// ```
    #[must_use]
    pub const fn should_display(&self) -> bool {
        !matches!(self, Self::LogOnly)
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_type_to_log_level() {
        assert_eq!(MessageType::Debug.to_log_level(), log::Level::Debug);
        assert_eq!(MessageType::Info.to_log_level(), log::Level::Info);
        assert_eq!(MessageType::Warning.to_log_level(), log::Level::Warn);
        assert_eq!(MessageType::Error.to_log_level(), log::Level::Error);
        assert_eq!(MessageType::Success.to_log_level(), log::Level::Info);
        assert_eq!(MessageType::Progress.to_log_level(), log::Level::Debug);
        assert_eq!(MessageType::Critical.to_log_level(), log::Level::Error);
    }

    #[test]
    fn test_message_type_name() {
        assert_eq!(MessageType::Info.name(), "Info");
        assert_eq!(MessageType::Warning.name(), "Warning");
        assert_eq!(MessageType::Error.name(), "Error");
        assert_eq!(MessageType::Success.name(), "Success");
        assert_eq!(MessageType::Progress.name(), "Progress");
        assert_eq!(MessageType::Debug.name(), "Debug");
        assert_eq!(MessageType::Critical.name(), "Critical");
    }

    #[test]
    fn test_message_target_should_display_in_gui() {
        assert!(MessageTarget::All.should_display_in_gui());
        assert!(MessageTarget::GuiOnly.should_display_in_gui());
        assert!(MessageTarget::Gui.should_display_in_gui());
        assert!(!MessageTarget::CliOnly.should_display_in_gui());
        assert!(!MessageTarget::Console.should_display_in_gui());
        assert!(!MessageTarget::LogOnly.should_display_in_gui());
    }

    #[test]
    fn test_message_target_should_display_in_cli() {
        assert!(MessageTarget::All.should_display_in_cli());
        assert!(MessageTarget::CliOnly.should_display_in_cli());
        assert!(MessageTarget::Console.should_display_in_cli());
        assert!(!MessageTarget::GuiOnly.should_display_in_cli());
        assert!(!MessageTarget::Gui.should_display_in_cli());
        assert!(!MessageTarget::LogOnly.should_display_in_cli());
    }

    #[test]
    fn test_message_target_should_display() {
        assert!(MessageTarget::All.should_display());
        assert!(MessageTarget::GuiOnly.should_display());
        assert!(MessageTarget::CliOnly.should_display());
        assert!(MessageTarget::Gui.should_display());
        assert!(MessageTarget::Console.should_display());
        assert!(!MessageTarget::LogOnly.should_display());
    }

    #[test]
    fn test_message_target_default() {
        assert_eq!(MessageTarget::default(), MessageTarget::All);
    }
}

File: ClassicLib/MessageHandler/core/enums.py
"""Message type and target enums for the MessageHandler system.

This module provides enumerations for message categorization and routing.
When Rust acceleration is available, these can interoperate with the
Rust enums via the classic_message module.

Attributes:
    RUST_ENUMS: Whether Rust enum integration is available.
"""

from __future__ import annotations

from enum import Enum, auto
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from typing import Self

# Try to import Rust enums for interop
try:
    from classic_message import MessageTarget as RustMessageTarget
    from classic_message import MessageType as RustMessageType

    _RUST_ENUMS_AVAILABLE = True
except ImportError:
    _RUST_ENUMS_AVAILABLE = False
    RustMessageType = None  # type: ignore[assignment, misc]
    RustMessageTarget = None  # type: ignore[assignment, misc]

# Export as module constant
RUST_ENUMS: bool = _RUST_ENUMS_AVAILABLE


class MessageType(Enum):
    """Message type enumeration for categorizing messages.

    Attributes:
        INFO: Informational message.
        WARNING: Warning message indicating a potential issue.
        ERROR: Error message indicating a failure or problem.
        SUCCESS: Message indicating successful completion.
        PROGRESS: Message signaling progress of an ongoing operation.
        DEBUG: Message intended for debugging purposes.
        CRITICAL: Message indicating a critical issue requiring attention.
    """

    INFO = auto()
    WARNING = auto()
    ERROR = auto()
    SUCCESS = auto()
    PROGRESS = auto()
    DEBUG = auto()
    CRITICAL = auto()

    def to_rust(self) -> Any | None:
        """Convert to Rust MessageType if available.

        Returns:
            Rust MessageType enum value or None if Rust unavailable.
        """
        if not RUST_ENUMS or RustMessageType is None:
            return None

        mapping = {
            MessageType.INFO: RustMessageType.Info,
            MessageType.WARNING: RustMessageType.Warning,
            MessageType.ERROR: RustMessageType.Error,
            MessageType.SUCCESS: RustMessageType.Success,
            MessageType.PROGRESS: RustMessageType.Progress,
            MessageType.DEBUG: RustMessageType.Debug,
            MessageType.CRITICAL: RustMessageType.Critical,
        }
        return mapping.get(self)

    @classmethod
    def from_rust(cls, rust_type: Any) -> Self:
        """Convert from Rust MessageType.

        Args:
            rust_type: Rust MessageType enum value.

        Returns:
            Corresponding Python MessageType.

        Raises:
            ValueError: If unknown Rust type.
        """
        if not RUST_ENUMS:
            msg = "Rust enums not available"
            raise ValueError(msg)

        # Use the name for mapping
        name = str(rust_type).lower()
        for member in cls:
            if member.name.lower() == name:
                return member

        msg = f"Unknown Rust MessageType: {rust_type}"
        raise ValueError(msg)


class MessageTarget(Enum):
    """Message target enumeration for routing messages.

    Attributes:
        ALL: Show in both GUI and CLI.
        GUI: Show only in GUI mode.
        CONSOLE: Show only in CLI mode.
        LOG_ONLY: Only write to log file, no display.
    """

    ALL = auto()
    GUI = auto()
    CONSOLE = auto()
    LOG_ONLY = auto()

    def normalize(self) -> MessageTarget:
        """Return self (no normalization needed with canonical values only).

        Returns:
            The same MessageTarget value.
        """
        return self

    def should_display_in_gui(self) -> bool:
        """Check if message should display in GUI mode.

        Returns:
            True if should display in GUI.
        """
        normalized = self.normalize()
        return normalized in {MessageTarget.ALL, MessageTarget.GUI}

    def should_display_in_cli(self) -> bool:
        """Check if message should display in CLI mode.

        Returns:
            True if should display in CLI.
        """
        normalized = self.normalize()
        return normalized in {MessageTarget.ALL, MessageTarget.CONSOLE}

    def should_display(self) -> bool:
        """Check if message should display at all (not log-only).

        Returns:
            True if should display.
        """
        normalized = self.normalize()
        return normalized != MessageTarget.LOG_ONLY

    def to_rust(self) -> Any | None:
        """Convert to Rust MessageTarget if available.

        Returns:
            Rust MessageTarget enum value or None if Rust unavailable.
        """
        if not RUST_ENUMS or RustMessageTarget is None:
            return None

        # Normalize first, then map
        normalized = self.normalize()
        mapping = {
            MessageTarget.ALL: RustMessageTarget.All,
            MessageTarget.GUI: RustMessageTarget.Gui,
            MessageTarget.CONSOLE: RustMessageTarget.Console,
            MessageTarget.LOG_ONLY: RustMessageTarget.LogOnly,
        }
        return mapping.get(normalized)

    @classmethod
    def from_rust(cls, rust_target: Any) -> MessageTarget:
        """Convert from Rust MessageTarget.

        Args:
            rust_target: Rust MessageTarget enum value.

        Returns:
            Corresponding Python MessageTarget.

        Raises:
            ValueError: If unknown Rust target.
        """
        if not RUST_ENUMS:
            msg = "Rust enums not available"
            raise ValueError(msg)

        # Use string representation for mapping
        name = str(rust_target).lower()
        # Map to canonical names
        if "gui" in name:
            return cls.GUI
        if "console" in name or "cli" in name:
            return cls.CONSOLE
        if "all" in name:
            return cls.ALL
        if "log" in name:
            return cls.LOG_ONLY

        msg = f"Unknown Rust MessageTarget: {rust_target}"
        raise ValueError(msg)
